File: sdk/nodejs/src/http-client.js
'use strict';
const axios = require('axios');

class HTTPClient {
  constructor({ baseUrl = 'https://api-tools.hitechcloud.vn', apiKey, timeout = 30000, maxRetries = 3 } = {}) {
    this.baseUrl = baseUrl.replace(/\/$/, '');
    this.maxRetries = maxRetries;
    this.client = axios.create({
      baseURL: this.baseUrl,
      timeout,
      headers: { 'Accept': 'application/json', 'User-Agent': 'HiTechCloud-Tools-Node-SDK/1.0.0' },
    });
    if (apiKey) this.setApiKey(apiKey);
  }

  setApiKey(key) { this.client.defaults.headers.common['X-API-Key'] = key; }
  clearApiKey() { delete this.client.defaults.headers.common['X-API-Key']; }

  async request(method, path, { params, data } = {}) {
    let lastError;
    for (let attempt = 0; attempt < this.maxRetries; attempt++) {
      try {
        const resp = await this.client.request({ method, url: path, params, data });
        return resp.data;
      } catch (err) {
        if (err.response && err.response.status === 429) {
          const retryAfter = parseInt(err.response.headers['retry-after'] || '60', 10);
          if (attempt < this.maxRetries - 1) { await new Promise(r => setTimeout(r, retryAfter * 1000)); continue; }
        }
        if (err.response && err.response.status >= 400) {
          const body = err.response.data;
          const msg = body?.error || body?.message || `HTTP ${err.response.status}`;
          throw new Error(Array.isArray(msg) ? msg.join('; ') : msg);
        }
        lastError = err;
        if (attempt < this.maxRetries - 1) { await new Promise(r => setTimeout(r, 2 ** attempt * 1000)); continue; }
      }
    }
    throw lastError || new Error('Request failed');
  }

  get(path, params) { return this.request('GET', path, { params }); }
  post(path, data) { return this.request('POST', path, { data }); }
  close() {}
}

module.exports = { HTTPClient };

File: sdk/r/R/http_client.R
#' HTTP Client for HiTechCloud API
#' @param base_url API base URL
#' @param api_key = NULL
#' @param timeout Request timeout in seconds
#' @param max_retries Maximum retry attempts
#' @return HttpClient object
#' @export
HttpClient <- function(base_url = "https://api-tools.hitechcloud.vn", api_key = NULL, timeout = 30, max_retries = 3) {
  env <- new.env(parent = emptyenv())
  env$base_url <- base_url
  env$api_key <- api_key
  env$timeout <- timeout
  env$max_retries <- max_retries

  env$get <- function(path, params = NULL) {
    req <- request(paste0(env$base_url, path)) |>
      req_headers("Content-Type" = "application/json", "Accept" = "application/json") |>
      req_headers("X-API-Key" = env$api_key) |>
      req_timeout(env$timeout)
    if (!is.null(params)) req <- req |> req_url_query(!!!params)
    resp_body_json(req_perform(req))
  }

  env$post <- function(path, body = NULL) {
    req <- request(paste0(env$base_url, path)) |>
      req_headers("Content-Type" = "application/json", "Accept" = "application/json") |>
      req_headers("X-API-Key" = env$api_key) |>
      req_timeout(env$timeout) |>
      req_body_json(body %||% list())
    resp_body_json(req_perform(req))
  }

  env$put <- function(path, body = NULL) {
    req <- request(paste0(env$base_url, path)) |>
      req_headers("Content-Type" = "application/json", "Accept" = "application/json") |>
      req_headers("X-API-Key" = env$api_key) |>
      req_timeout(env$timeout) |>
      req_method("PUT") |>
      req_body_json(body %||% list())
    resp_body_json(req_perform(req))
  }

  env$delete <- function(path) {
    req <- request(paste0(env$base_url, path)) |>
      req_headers("Content-Type" = "application/json", "Accept" = "application/json") |>
      req_headers("X-API-Key" = env$api_key) |>
      req_timeout(env$timeout) |>
      req_method("DELETE")
    resp_body_json(req_perform(req))
  }

  class(env) <- "HttpClient"
  env
}
